folio-spring-i18n/src/main/java/org/folio/spring/i18n/model/TranslationMap.java

```diff
@@ -124,64 +121,103 @@ public String get(String key) {
-              .atDate(LocalDate.now())
-              .atZone(ZoneId.systemDefault())
-              .toInstant()
-          );
-      }
-    }
-    return new MessageFormat(get(key), locale).format(buildMap(args));
+  public String format(ZoneId zone, String key, Object... args) {
+    return formatRaw(zone, new MessageFormat(get(key)), buildArgs(zone, args));
   }
 
   /**
-   * Get a translation map from a set of (String, Object) pairs.
+   * Like {@link #format(ZoneId, String, Object...)}, but uses a message format string rather than
+   * looking it up in the map.
    *
-   * @param args pairs of elements, e.g. key1, value1, key2, value2, ...
-   * @return the map of each key =&gt; value
-   * @throws IllegalArgumentException if an odd number of parameters is passed
+   * @param zone the timezone to use for date formatting
+   * @param format the format string
+   * @param args pairs of keys and values to interpolate
+   * @return the formatted string
    */
-  private static Map<String, Object> buildMap(Object... args) {
+  public String formatString(ZoneId zone, String format, Object... args) {
+    return formatRaw(zone, new MessageFormat(format), buildArgs(zone, args));
+  }
+
+  /**
+   * Build associative map to pass to {@link MessageFormat#format(String, Map)}.
+   */
+  private static Map<String, Object> buildArgs(ZoneId zone, Object... args) {
     if (args.length % 2 != 0) {
       throw new IllegalArgumentException(
-        "An odd number of parameters were passed to buildMap; even amounts are needed to construct key-value pairs"
+        "An odd number of arguments were passed to buildArgs; even amounts are needed to construct key-value pairs"
       );
     }
 
     Map<String, Object> map = new HashMap<>();
 
     for (int i = 0; i < args.length; i += 2) {
+      // Sadly, ICU formatting strings only support date formats with the old Date class :(
+
+      if (args[i + 1] instanceof Instant instant) {
+        args[i + 1] = Date.from(instant);
+      }
+      if (args[i + 1] instanceof LocalDateTime date) {
+        args[i + 1] = Date.from(date.atZone(zone).toInstant());
+      }
+      if (args[i + 1] instanceof OffsetDateTime date) {
+        args[i + 1] = Date.from(date.toInstant());
+      }
+      if (args[i + 1] instanceof ZonedDateTime date) {
+        args[i + 1] = Date.from(date.toInstant());
+      }
+
+      // ICU will chop off the time, so the time we use is irrelevant
+      if (args[i + 1] instanceof LocalDate date) {
+        args[i + 1] = Date.from(date.atStartOfDay(zone).toInstant());
+      }
+
+      // ICU will chop off the date, so the date we use is irrelevant
+      if (args[i + 1] instanceof LocalTime time) {
+        args[i + 1] = Date.from(time.atDate(LocalDate.now()).atZone(zone).toInstant());
+      }
+      if (args[i + 1] instanceof OffsetTime time) {
+        args[i + 1] = Date.from(time.atDate(LocalDate.now()).toInstant());
+      }
+
       map.put(args[i].toString(), args[i + 1]);
```

Comment:
I consolidated them into a pattern matching switch, a preview feature in JDK 17. The feature ([JEP 441](https://openjdk.org/jeps/441)) is standard as of JDK 21, and there are no API changes between 17 and then which would break this usage, so we are safe to use it now.

If this is unacceptable, I can turn it back into an else-if, but I felt this was the best solution.